component/common/loki/wal/watcher.go

```diff
@@ -331,9 +349,24 @@ func (w *Watcher) decodeAndDispatch(b []byte, segmentNum int) (bool, error) {
 	return readData, firstErr
 }
 
-func (w *Watcher) Stop() {
-	// first close the quit channel to order main mainLoop routine to stop
-	close(w.quit)
+// Stop stops the Watcher, draining the WAL until the end of the last segment if drain is true. Since the writer of the WAL
+// is expected to have stopped before the Watcher, no further writes are expected, so segments can be safely consumed.
+//
+// Note if drain is enabled, the caller routine of Stop will block executing the drain procedure.
+func (w *Watcher) Stop(drain bool) {
```

Comment:
Part of me think this is combining two state transitions in one call which might be confusing?

Reply:
Mmm actually you right, it's almost identical to split apart into a `Stop` and `Drain` methods, let me try that

Reply:
Ok splitted apart into `Stop` and `Drain`